x/mongo/driver/topology/server.go

```diff
@@ -437,119 +526,157 @@ func (s *Server) updateDescription(desc description.Server) {
+		return nil
+	}))
 
-	for i := 1; i <= maxRetry; i++ {
-		var now time.Time
-		var descPtr *description.Server
+	return newConnection(s.address, opts...)
+}
 
-		if conn != nil && conn.expired() {
-			if conn.nc != nil {
-				conn.nc.Close()
-			}
-			conn = nil
-		}
+func (s *Server) setupHeartbeatConnection() error {
+	conn, err := s.createConnection()
+	if err != nil {
+		return err
+	}
 
-		if conn == nil {
-			opts := []ConnectionOption{
-				WithConnectTimeout(func(time.Duration) time.Duration { return s.cfg.heartbeatTimeout }),
-				WithReadTimeout(func(time.Duration) time.Duration { return s.cfg.heartbeatTimeout }),
-				WithWriteTimeout(func(time.Duration) time.Duration { return s.cfg.heartbeatTimeout }),
-			}
-			opts = append(opts, s.cfg.connectionOpts...)
-			// We override whatever handshaker is currently attached to the options with a basic
-			// one because need to make sure we don't do auth.
-			opts = append(opts, WithHandshaker(func(h Handshaker) Handshaker {
-				now = time.Now()
-				return operation.NewIsMaster().AppName(s.cfg.appname).Compressors(s.cfg.compressionOpts)
-			}))
-
-			// Override any command monitors specified in options with nil to avoid monitoring heartbeats.
-			opts = append(opts, WithMonitor(func(*event.CommandMonitor) *event.CommandMonitor {
-				return nil
-			}))
-
-			conn, err = newConnection(ctx, s.address, opts...)
-
-			conn.connect(ctx)
-
-			err = conn.wait()
-			if err == nil {
-				descPtr = &conn.desc
-			}
-		}
+	// Take the lock when assigning the context and connection because they're accessed by cancelCheck.
+	s.heartbeatLock.Lock()
+	s.heartbeatCtx, s.heartbeatCtxCancel = context.WithCancel(s.globalCtx)
+	s.conn = conn
+	s.heartbeatLock.Unlock()
 
-		// do a heartbeat because a new connection wasn't created so a handshake was not performed
-		if descPtr == nil && err == nil {
-			now = time.Now()
-			op := operation.
-				NewIsMaster().
-				ClusterClock(s.cfg.clock).
-				Deployment(driver.SingleConnectionDeployment{initConnection{conn}})
-			err = op.Execute(ctx)
-			if err == nil {
-				tmpDesc := op.Result(s.address)
-				descPtr = &tmpDesc
-			} else {
-				// close the connection here rather than in the error check below to avoid calling Close on a net.Conn
-				// that wasn't successfully created
-				_ = conn.close()
-			}
+	s.conn.connect(s.heartbeatCtx)
+	return s.conn.wait()
+}
+
+// cancelCheck cancels in-progress connection dials and reads. It does not set any fields on the server.
+func (s *Server) cancelCheck() {
+	var conn *connection
+
+	// Take heartbeatLock for mutual exclusion with the checks in the update function.
+	s.heartbeatLock.Lock()
+	if s.heartbeatCtx != nil {
+		s.heartbeatCtxCancel()
+	}
+	conn = s.conn
+	s.heartbeatLock.Unlock()
+
+	if conn == nil {
+		return
+	}
+
+	// If the connection exists, we need to wait for it to be connected. We can ignore the error from conn.wait(). If
```

Comment:
I think this comment would be more helpful if it explained _why_ we need to wait until it's connected. What do you think? Is it because we cannot call close() while the connection is being connected?